fix: implement `getblock` and `getblockheader` compliant with core

qlrd opened the pull request:
### What is the purpose of this pull request?

- [x] Bug fix
- [x] Documentation update
- [ ] New feature
- [x] Test
- [ ] Other: <!-- Please describe it -->

### Which crates are being modified?


- [ ] floresta-chain
- [x] floresta-cli
- [ ] floresta-common
- [ ] floresta-compact-filters
- [ ] floresta-electrum
- [x] floresta-node
- [x] floresta-rpc
- [ ] floresta-watch-only
- [ ] floresta-wire
- [ ] floresta
- [ ] florestad
- [ ] Other: <!-- Please describe it -->

### Description and Notes

Fix #655

### How to verify the changes you have done?

Add a `verbose` paramenter to `getblockheader` command so it will comply with core wheather to return a "raw" blockheader (as a hexadecimal string) or a "verbose" blockheader (as a json object).

Due to this, it was also needed to fix `getblock` rpc method so it can, now, use verbosity levels `0` up to `2`.

qlrd committed:
fix: implement `getblock`and `getblockheader` compliant with core

This commit add a `verbose` paramenter to `getblockheader` command so it
will comply with core wheather to return a "raw" blockheader (as a
hexadecimal string) or a "verbose" blockheader (as a json object). It
also add a unit test for `getblockheader` in both raw and verbose mode,
that must result in a proper
`serde_json::Value<GetBlockHeaderResVerbose>`.

Fix #655.

fix

qlrd committed:
test: apply integration tests to the fixed `getblockheader`and `getblock` rpc

This commit add a cycle of connections and mining to the
`getblockheader`/`getblock` tesst so it can check the blocks/
blockheaders for 102 different heights.

qlrd committed:
docs: add documentation to `getblockheader` rpc

jaoleal commented on bin/floresta-cli/src/main.rs:
```diff
@@ -223,9 +223,16 @@ pub enum Methods {
+    )]
+    GetBlockHeader {
+        hash: BlockHash,
+        verbose: Option<bool>,
```
What about making this a proper clap flag ? An example is our `use_timestamp` from RescanBlockchain

qlrd replied:
Thanks @jaoleal, will take a look on that

jaoleal reviewed (changes requested):
I know that this is draft and the CI is broken but heres some early thougths

Davidson-Souza commented on crates/floresta-node/src/json_rpc/server.rs:
```diff
@@ -232,6 +237,163 @@ impl<Blockchain: RpcChain> RpcImpl<Blockchain> {
             .map_err(|_| JsonRpcError::Node("Failed to get peer info".to_string()))
     }
+
+    async fn get_blockheader(
```
35aa245c8bb50b6de099be547c3a8f3cb6b55abd

This function should be in `blockchain.rs`

Davidson-Souza commented on bin/floresta-cli/src/main.rs:
```diff
@@ -223,9 +224,21 @@ pub enum Methods {
-    GetBlockHeader { hash: BlockHash },
+    #[command(
+        name = "getblockheader",
+        about = "Returns a string that is serialized, hex-encoded data for blockheader ‘hash’ when `verbose` is false, else returns a json object.",
```
Returns the header for a block with a given `hash`, with optional information about it

Davidson-Souza commented on crates/floresta-node/src/json_rpc/blockchain.rs:
```diff
@@ -327,10 +357,134 @@ impl<Blockchain: RpcChain> RpcImpl<Blockchain> {
+                .to_lower_hex_string(),
+            target: Target::from_compact(header.bits)
+                .to_be_bytes()
+                .to_lower_hex_string(),
```
If you don't get the bytes, will it serialize wrong?

Davidson-Souza commented on crates/floresta-node/src/json_rpc/res.rs:
```diff
@@ -310,6 +366,15 @@ pub enum JsonRpcError {
+
+    InvalidBlockHash(Box<dyn std::error::Error + Send + Sync>),
+
+    InvalidHeader(Box<dyn std::error::Error + Send + Sync + 'static>),
```
I fell like they are basically `BlockNotFound`. Also, no need to use these `Box`-ed errors

Davidson-Souza commented on tests/test_framework/__init__.py:
```diff
@@ -852,3 +852,13 @@ def assertHasAny(self, actual: Any, pattern: Pattern) -> None:
+            self.stop()
+            raise AssertionError(
+                f"Actual: {actual} ~ {expected}\n Expected: {actual} ~ {expected}"
+            )
```
Nice trick, this is clever 😃

Davidson-Souza commented on doc/rpc/getblockheader.md:
```diff
@@ -0,0 +1,79 @@
+  "nTx" : n,                      (numeric) The number of transactions in the block
+  "previousblockhash" : "hex",    (string) The hash of the previous block
+  "nextblockhash" : "hex"         (string) The hash of the next block
+}
```
We went for a different formatting here. See the other files for more info

moisesPompilio commented:
Need rebase

qlrd committed:
chore: remove unused imports and add ignore files

This commit remove unused imports in `tests/test_framework/__init__.py`
as well add a ignore pattern to .gitignore for macOS `.DS_STORE` files.

lorenzolfm commented:
Did you picked this up after all, @Micah-Shallom? I can review the PR if its still open

Micah-Shallom commented:
Hi @lorenzolfm ...its currently a work in progress in my local branch...will raise it soon enough
